applications/machine_learning/src/modules/led_state.c

```diff
@@ -10,45 +10,255 @@
+		const struct ml_result_led_effect *t = &ml_result_led_effects[i];
+
+		if (!strcmp(t->label, label)) {
+			result = t;
+			break;
+		}
+	}
+
+	return result;
+}
+
+static void display_sensor_sim(const char *label)
+{
+	static const struct ml_result_led_effect *sensor_sim_effect;
+
+	if (label) {
+		sensor_sim_effect = get_led_effect(label);
+
+		if (sensor_sim_effect == DEFAULT_EFFECT) {
+			LOG_WRN("No LED effect for sensor_sim label %s", label);
+		}
+	}
+
+	if (sensor_sim_effect) {
+		__ASSERT_NO_MSG(sensor_sim_effect->effect.loop_forever);
+		send_led_event(led_map[LED_ID_SENSOR_SIM], &sensor_sim_effect->effect);
+	}
+}
+
+static void display_ml_result(const char *label, bool force_update)
+{
+	__ASSERT_NO_MSG(ml_state == ML_STATE_MODEL_RUNNING);
+
+	static const struct ml_result_led_effect *ml_result_effect;
+	const struct ml_result_led_effect *new_effect = get_led_effect(label);
+
+	/* Update not needed. */
+	if ((ml_result_effect == new_effect) && !force_update) {
+		return;
+	}
+
+	__ASSERT_NO_MSG(!force_update || !label);
+
+	if (!force_update) {
+		if (!label) {
+			LOG_INF("Anomaly detected");
+		} else if (new_effect == DEFAULT_EFFECT) {
+			LOG_INF("No LED effect for label: %s", label);
+		} else {
+			LOG_INF("Displaying LED effect for label: %s", label);
+		}
+	}
+
+	/* Synchronize LED effect displayed for simulated signal. */
+	if (DISPLAY_SIM_SIGNAL && (new_effect != DEFAULT_EFFECT)) {
+		display_sensor_sim(NULL);
+	}
+
+	ml_result_effect = new_effect;
+	send_led_event(led_map[LED_ID_ML_STATE], &ml_result_effect->effect);
+
+	if (!ml_result_effect->effect.loop_forever) {
+		blocking_led_effect = &ml_result_effect->effect;
+	} else {
+		blocking_led_effect = NULL;
+	}
+}
+
+static void update_ml_result(const char *label, float value, float anomaly)
+{
+	const char *new_label = cur_label;
+	bool increment = true;
+
+	if (anomaly > ANOMALY_THRESH) {
+		new_label = NULL;
+	} else if (value >= VALUE_THRESH) {
+		new_label = label;
+	} else {
+		increment = false;
+	}
+
+	if (new_label != cur_label) {
+		if ((!new_label || !cur_label) || strcmp(new_label, cur_label)) {
+			cur_label = new_label;
+			prediction_streak = 0;
+		}
+	}
+
+	if (increment) {
+		prediction_streak++;
+	}
+
+	if (prediction_streak >= PREDICTION_STREAK_THRESH) {
+		display_ml_result(cur_label, false);
+		clear_prediction();
+	}
+}
+
+static void validate_configuration(void)
+{
+	BUILD_ASSERT(ARRAY_SIZE(ml_result_led_effects) >= 1);
+	__ASSERT_NO_MSG(DEFAULT_EFFECT->effect.loop_forever);
+	__ASSERT_NO_MSG(!DEFAULT_EFFECT->label);
+
+	for (size_t i = 1; i < ARRAY_SIZE(ml_result_led_effects); i++) {
+		const struct ml_result_led_effect *t = &ml_result_led_effects[i];
+		const uint8_t zeros[sizeof(t->effect)] = {0};
+
+		__ASSERT_NO_MSG(memcmp(&t->effect, zeros, sizeof(t->effect)));
+		__ASSERT_NO_MSG(t->label);
+	}
+}
+
+static bool handle_ml_result_event(const struct ml_result_event *event)
+{
+	if ((ml_state == ML_STATE_MODEL_RUNNING) && !blocking_led_effect) {
```

Comment:
regarding `!blocking_led_effect`; shouldn't we record the ml_result and apply it on the first chance possible?
I guess you depend on the result being repeated but will it always be the case? won' the possible delay be too long?

Reply:
My preferred approach was to "unsubscribe" while result is displayed to make a demo for the ml_runner subscription mechanism. Blocking led effects also makes prediction start synchronized with displayed effects (user knows when the machine learning was restarted).